datafusion/core/src/datasource/datasource.rs

```diff
@@ -72,3 +72,12 @@ pub trait TableProvider: Sync + Send {
         Ok(TableProviderFilterPushDown::Unsupported)
     }
 }
+
+/// A factory which creates [`TableProvider`]s at runtime given a URL.
+///
+/// For example, this can be used to create a table "on the fly"
+/// from a directory of files only when that name is referenced.  
+pub trait TableProviderFactory: Sync + Send {
```

Comment:
Should `create` be `async`? 

The `TableProvider` will have to know the `Schema` which will most likely have to be inferred by reading from the `ObjectStore` (or involve a network call otherwise if there is some sort of metastore involved).